lib/puppet/pops/types/types.rb

```diff
@@ -1,415 +1,1482 @@
+        callable_t.assignable?(DEFAULT)
+      end
+
+      DEFAULT = PUndefType.new
+
+      protected
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PUndefType)
       end
     end
 
+    # A type private to the type system that describes "ignored type" - i.e. "I am what you are"
+    # @api private
+    #
+    class PUnitType < PAnyType
+      def instance?(o)
+        true
+      end
+
+      DEFAULT = PUnitType.new
+
+      protected
+      # @api private
+      def _assignable?(o)
+        true
+      end
+    end
+
+    # @api public
+    #
+    class PDefaultType < PAnyType
+      def instance?(o)
+        o == :default
+      end
+
+      DEFAULT = PDefaultType.new
+
+      protected
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PDefaultType)
+      end
+    end
+
+    # A flexible data type, being assignable to its subtypes as well as PArrayType and PHashType with element type assignable to PDataType.
+    #
+    # @api public
+    #
     class PDataType < PAnyType
-      module ClassModule
-        def ==(o)
-          self.class == o.class ||
-            o.class == PVariantType && o == Puppet::Pops::Types::TypeCalculator.data_variant()
+      def ==(o)
+        self.class == o.class || o.class == PVariantType && o == PVariantType::DATA
+      end
+
+      def instance?(o)
+        PVariantType::DATA.instance?(o)
+      end
+
+      DEFAULT = PDataType.new
+
+      protected
+
+      # Data is assignable by other Data and by Array[Data] and Hash[Scalar, Data]
+      # @api private
+      def _assignable?(o)
+        # We cannot put the NotUndefType[Data] in the @data_variant_t since that causes an endless recursion
+        case o
+        when Types::PDataType
+          true
+        when Types::PNotUndefType
+          assignable?(o.type || PUndefType::DEFAULT)
+        else
+          PVariantType::DATA.assignable?(o)
         end
       end
     end
 
-    class PVariantType < PAnyType
-      module ClassModule
+    # Type that is PDataType compatible, but is not a PCollectionType.
+    # @api public
+    #
+    class PScalarType < PAnyType
 
-        def hash
-          [self.class, Set.new(self.types)].hash
-        end
+      def instance?(o)
+        assignable?(TypeCalculator.infer(o))
+      end
 
-        def ==(o)
-          (self.class == o.class && Set.new(types) == Set.new(o.types)) ||
-            (o.class == PDataType && self == Puppet::Pops::Types::TypeCalculator.data_variant())
-        end
+      DEFAULT = PScalarType.new
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PScalarType)
       end
     end
 
+    # A string type describing the set of strings having one of the given values
+    # @api public
+    #
     class PEnumType < PScalarType
-      module ClassModule
-        def hash
-          [self.class, Set.new(self.values)].hash
+      attr_reader :values
+
+      def initialize(values)
+        @values = values.sort.freeze
```

Comment:
this sort makes the source and string form of the type differ - it also makes a difference if an enum is mapped to numeric form as additions can be placed in the middle. I think we an live with those "problems" though.